exporters/otlp/trace/src/main/java/io/opentelemetry/exporter/otlp/trace/OtlpHttpUtil.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.exporter.otlp.trace;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.io.SegmentedStringWriter;
+import com.google.protobuf.util.JsonFormat;
+import com.google.rpc.Status;
+import io.opentelemetry.exporter.otlp.internal.HexEncodingStringJsonGenerator;
+import io.opentelemetry.exporter.otlp.trace.OtlpHttpSpanExporter.RequestResponseHandler;
+import io.opentelemetry.proto.collector.trace.v1.ExportTraceServiceRequest;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import okhttp3.Interceptor;
+import okhttp3.MediaType;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+import okhttp3.tls.HandshakeCertificates;
+import okio.BufferedSink;
+import okio.GzipSink;
+import okio.Okio;
+import org.curioswitch.common.protobuf.json.MessageMarshaller;
+
+final class OtlpHttpUtil {
+
+  static final RequestResponseHandler PROTOBUF_REQUEST_RESPONSE_HANDLER =
+      new ProtobufRequestResponseHandler();
+  static final RequestResponseHandler JSON_REQUEST_RESPONSE_HANDLER =
+      new JsonRequestResponseHandler();
+  static final Interceptor GZIP_INTERCEPTOR = new GzipInterceptor();
+
+  /**
+   * Extract X.509 certificates from the bytes.
+   *
+   * @param trustedCertificatesPem bytes containing an X.509 certificate collection in PEM format.
+   * @return a HandshakeCertificates with the certificates
+   * @throws CertificateException if an error occurs extracting certificates
+   */
+  static HandshakeCertificates toHandshakeCertificates(byte[] trustedCertificatesPem)
+      throws CertificateException {
+    InputStream is = new ByteArrayInputStream(trustedCertificatesPem);
+    CertificateFactory factory = CertificateFactory.getInstance("X.509");
+    HandshakeCertificates.Builder certBuilder = new HandshakeCertificates.Builder();
+    X509Certificate cert;
+    while ((cert = (X509Certificate) factory.generateCertificate(is)) != null) {
+      certBuilder.addTrustedCertificate(cert);
+    }
+    return certBuilder.build();
+  }
+
+  private static class ProtobufRequestResponseHandler implements RequestResponseHandler {
+
+    @Override
+    public RequestBody build(ExportTraceServiceRequest exportTraceServiceRequest) {
+      return RequestBody.create(
+          exportTraceServiceRequest.toByteArray(), MediaType.parse("application/x-protobuf"));
```

Comment:
It can and probably should be a future PR but would definitely be good to use okio buffer and codedinputstream to skip the intermediate unpooled buffer